tests/test_pycrdt_yjs.py

```diff
@@ -1,75 +1,95 @@
 
-        subscription_id = self.ytest.observe(callback)
-        async with create_task_group():
-            with move_on_after(self.timeout):
-                await change.wait()
-
-        self.ytest.unobserve(subscription_id)
+        subscription_id = self.ymap.observe(callback)
+        # wait up to 1000ms for a reply
+        with move_on_after(1.0):
+            await change.wait()
+        self.ymap.unobserve(subscription_id)
 
 
 @pytest.mark.anyio
 @pytest.mark.parametrize("yjs_client", "0", indirect=True)
 async def test_pycrdt_yjs_0(yws_server, yjs_client):
+    """
+    The JS client (`yjs_client_0.js`) should set `ymap["out"] := ymap["in"] + 1`
+    whenever the clock is incremented via `yclient.inc_clock()`.
+    """
+    MSG_COUNT = 10
     ydoc = Doc()
-    ytest = YTest(ydoc)
+    yclient = YClient(ydoc)
+    ymap = yclient.ymap
+
     async with connect("ws://127.0.0.1:1234/my-roomname") as websocket, WebsocketProvider(
         ydoc, websocket
     ):
-        ydoc["map"] = ymap = Map()
-        # set a value in "in"
-        for v_in in range(10):
-            ymap["in"] = float(v_in)
-            ytest.run_clock()
-            await ytest.clock_run()
-            v_out = ymap["out"]
-            assert v_out == v_in + 1.0
+        for msg_num in range(1, MSG_COUNT + 1):
+            # set `ymap["in"]`
+            ymap["in"] = float(msg_num)
+            yclient.inc_clock()
+            await yclient.wait_for_reply()
+
+            # assert JS client increments `ymap["out"]` in response
+            assert ymap["out"] == ymap["in"] + 1.0
+
+    assert yclient.clock == MSG_COUNT * 2
 
 
 @pytest.mark.anyio
 @pytest.mark.parametrize("yjs_client", "1", indirect=True)
 async def test_pycrdt_yjs_1(yws_server, yjs_client):
-    # wait for the JS client to connect
-    tt, dt = 0, 0.1
-    while True:
-        await sleep(dt)
-        if "/my-roomname" in yws_server.rooms:
-            break
-        tt += dt
-        if tt >= 1:
-            raise RuntimeError("Timeout waiting for client to connect")
-    ydoc = yws_server.rooms["/my-roomname"].ydoc
-    ytest = YTest(ydoc)
-    ytest.run_clock()
-    await ytest.clock_run()
-    ydoc["cells"] = ycells = Array()
-    ydoc["state"] = ystate = Map()
-    assert ycells.to_py() == [{"metadata": {"foo": "bar"}, "source": "1 + 2"}]
-    assert ystate.to_py() == {"state": {"dirty": False}}
+    """
+    The JS client (`yjs_client_1.js`) should set `ydoc["cells"]
+    whenever the clock is incremented via `yclient.inc_clock()`.
+    """
+    yroom = await yws_server.get_room("/my-roomname")
+    ydoc = yroom.ydoc
+    yclient = YClient(ydoc)
+
+    yclient.inc_clock()
+    await yclient.wait_for_reply()
+
+    # TODO: remove the need to set a root type before accessing it
+    ydoc["cells"] = Array()
+    ydoc["state"] = Map()
+    assert ydoc["cells"].to_py() == [{"metadata": {"foo": "bar"}, "source": "1 + 2"}]
+    assert ydoc["state"].to_py() == {"state": {"dirty": False}}
```

Comment:
@davidbrochart I noticed that I had to bind an empty shared type to Ydoc keys before I am able to access their values. This test fails if you remove the definition statements. In Yjs however, you can just run `ydoc.get("<root-type-name>")` to get a reference to a shared type by its name, without having to call some equivalent of `ydoc.set(...)` beforehand.

Is that feature possible with `pycrdt`? If not, can we open an issue and plan to implement this?

Reply:
In Yjs you can do e.g. `ydoc.get("my_array", Y.Array)` which is the equivalent of `ydoc.getArray("my_array")`, which defines a top-level shared type. The equivalent in pycrdt is `ydoc["my_array"] = Array()`, so it's basically the same, just a different API.

Reply:
There are two drawbacks to only supporting this way of accessing a shared type:

1. This requires 1 additional line of code per shared type for the assignment statement, and can get verbose if one is using many. However, in Yjs, `ydoc.getArray(...)` can be used inline.

- The inline assignment operator `:=` (available in Python 3.8+) does not work either:

```
/Users/dlq/micromamba/envs/rtcdev/lib/python3.11/ast.py:50: in parse
    return compile(source, filename, mode, flags,
E     File "/Volumes/workplace/pycrdt-websocket/tests/test_pycrdt_yjs.py", line 92
E       assert (ydoc["cells"] := Array()).to_py() == [{"metadata": {"foo": "bar"}, "source": "1 + 2"}]
E               ^^^^^^^^^^^^^
E   SyntaxError: cannot use assignment expressions with subscript
```

2. The syntax requires assigning an empty shared type in order to access an existing, non-empty shared type. This is stateful and confusing, because the value of `ydoc["cells"]` is not the value of `ydoc["cells"]` that was just set in the immediately preceding line. This generally violates how most programming languages work. I understand that this is permitted by Python, but the public API should not rely on exotic behavior exclusive to Python.


@davidbrochart Given these two points, would you be open to accepting this as an issue on the `pycrdt` repo? I can create the issue if you lack the time.

Reply:
For a more Pythonic take, why not do something like `doc.get_array`? I will say that I think it is confusing to set a key to an empty object before reading a previously set non-empty value. I had to read the code many times and think carefully about what it was doing to understand it.